pkg/proxy/handlers/spacelister.go

```diff
@@ -0,0 +1,157 @@
+	userID, _ := ctx.Get(context.SubKey).(string)
+	username, _ := ctx.Get(context.UsernameKey).(string)
+	workspaceName := ctx.Param("workspace")
+	doGetWorkspace := len(workspaceName) > 0
+
+	signup, err := s.GetSignupFunc(userID, username)
+	if err != nil {
+		return err
+	}
+	if signup == nil || !signup.Status.Ready {
+		return errs.New("user is not provisioned (yet)")
+	}
+
+	murName := signup.CompliantUsername
+	murSelector, err := labels.NewRequirement(toolchainv1alpha1.SpaceBindingMasterUserRecordLabelKey, selection.Equals, []string{murName})
+	if err != nil {
+		return err
+	}
+
+	requirements := []labels.Requirement{*murSelector}
+
+	if doGetWorkspace {
+		// specific workspace requested so add label requirement to match the space
+		spaceSelector, err := labels.NewRequirement(toolchainv1alpha1.SpaceBindingSpaceLabelKey, selection.Equals, []string{workspaceName})
+		if err != nil {
+			return err
+		}
+		requirements = append(requirements, *spaceSelector)
+	}
+
+	spaceBindings, err := s.GetInformerServiceFunc().ListSpaceBindings(requirements...)
+	if err != nil {
+		return err
+	}
+
+	workspaces, err := s.workspacesFromSpaceBindings(spaceBindings)
+	if err != nil {
+		return err
+	}
+
+	if doGetWorkspace { // specific workspace requested
+		if len(workspaces) != 1 {
+			// not found
+			r := schema.GroupResource{Group: "toolchain.dev.openshift.com", Resource: "workspaces"}
+			return errorResponse(ctx, apierrors.NewNotFound(r, workspaceName), http.StatusNotFound)
+		}
+		return getWorkspaceResponse(ctx, workspaces[0])
+	}
+
+	return listWorkspaceResponse(ctx, workspaces)
+}
+
+func errorResponse(ctx echo.Context, err *apierrors.StatusError, code int) error {
+	ctx.Logger().Error(err, err.Error())
+	ctx.Response().Writer.Header().Set("Content-Type", "application/json")
+	ctx.Response().Writer.WriteHeader(code)
+	return json.NewEncoder(ctx.Response().Writer).Encode(err.ErrStatus)
+}
+
+func getWorkspaceResponse(ctx echo.Context, workspace toolchainv1alpha1.Workspace) error {
+	ctx.Response().Writer.Header().Set("Content-Type", "application/json")
+	ctx.Response().Writer.WriteHeader(http.StatusOK)
+	return json.NewEncoder(ctx.Response().Writer).Encode(workspace)
+}
+
+func listWorkspaceResponse(ctx echo.Context, workspaces []toolchainv1alpha1.Workspace) error {
+	workspaceList := &toolchainv1alpha1.WorkspaceList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "WorkspaceList",
+			APIVersion: "toolchain.dev.openshift.com/v1alpha1",
+		},
+		Items: workspaces,
+	}
+
+	ctx.Response().Writer.Header().Set("Content-Type", "application/json")
+	ctx.Response().Writer.WriteHeader(http.StatusOK)
+	return json.NewEncoder(ctx.Response().Writer).Encode(workspaceList)
+}
+
+func (s *SpaceLister) workspacesFromSpaceBindings(spaceBindings []*toolchainv1alpha1.SpaceBinding) ([]toolchainv1alpha1.Workspace, error) {
+	workspaces := []toolchainv1alpha1.Workspace{}
+	for _, spaceBinding := range spaceBindings {
+
+		if spaceBinding.Labels == nil {
```

Comment:
Labels can be nil? In case it's not fully initialized yet? Worth adding a comment and maybe log it..

Reply:
Not sure, figured I'd just cover it just in case (eg. if a SpaceBinding is created manually or something).

But actually now that I think of it it's redundant because `spaceName` would just be empty a few lines below and we would log the error when we try to find the space so it's covered. I removed this check in https://github.com/codeready-toolchain/registration-service/pull/301/commits/a1734586fc0770f0c714eb1a264d06ff9b3b349f
Thanks!